### ISSUE TITLE
Container mount directory in host leaks into another container namespace

### ISSUE THREAD

**Issue by rhvgoyal** [docker/docker#10991]
If I launch two containers as follows then mount directory of first container is available in second container too which in turn causes many other issues. I am using devicemapper driver.

- Launch first container, say A.
   $ docker run -ti fedora /bin/bash
   This will lead to mounting of thin device backing the container at /var/lib/docker/devicemapper/mnt/<container_id_A>

- Launch second container, say B.
  $ docker run -v /:/foo fedora /bin/bash

As we have bind mounted "/" of docker daemon namespace at /foo in container B namespace, 
container A's mount point is bind mounted too in container B at /foo/var/lib/docker/devicemapper/mnt/<container_id_A>.

And this creates multiple problems with container A as now one can not deactivate the thin device
backing container A or delete that thin device from pool.

- So if container A tries to exit, its exit is delayed by 10 seconds as we try to deactivate the device in a loop for 10 seconds. After 10 seonds, we leave a dangling device which will remain activated even after container B has exited.

- If one tries to remove container A, that fails as device is activated and it can't be removed from pool.
  But given current design of "docker --rm", we will remove container from docker data structures and
  container will not be visible to caller anymore but this leaks thin pool resources.

Can we unmount the container directory from host after container has been successfully launched. That way if somebody is launching another container with "-v /:/foo", it will not see the rootfs of first container and lot of problems should disappear.

Thoughts?

**Comment by rhvgoyal** [comment]
One more thing. Once the container has started and rootfs gets mounted inside container, unmounting it in host tries to deactivate thin device and that fails as now that device is being used from container too.

Device mapper provides a functionality where one can have deferred device removal. May be we can use that functionality for deactivating a device during unmount.

Hopefully it works fine with rest of the graph drivers.

**Comment by rhvgoyal** [comment]
I am about to post a patch series which will implement deferred device removal in devicemapper graph driver.  That will atleast take care of the issues when a container can not stop without delays as a busy device could not be removed.

**Comment by unclejack** [comment]
@crosbymichael What do you think about removing the container's directory from the host? It sounds reasonable.

@rhvgoyal Personally, I'd look into unmounting the container directory first and work on the deferred device removal in parallel.

**Comment by rhvgoyal** [comment]
@unclejack 

I already posted a patch series for unmounting container's directory here.

https://github.com/docker/docker/pull/11114

But I am not very happy with the solution now. Reason being that it assumes that once container is exiting there is no other user of device. But I want to keep a possibility open where container root could still remain mounted even after container is gone. So that some tools can inspect container. Or that directory could be bind mounted as volume inside another container for inspection etc.

Hence I started looking into deferred device removal solutions.

**Comment by rhvgoyal** [comment]
@vbatts 

Can we close this bug. Now deferred device removal patch series will take care of this partially. We might want to put in deferred device deletion capability too at some point of time. Can open a new issue whenever that happens. No point keeping it around till then.

**Comment by vbatts** [comment]
closed, at the op's request
